### DISCUSSION TITLE
Indexing with stake address

### DISCUSSION THREAD

**Discussion by baymac** [CardanoSolutions/kupo#48] [Archives]
Is it possible to get all transaction from addresses owned by stake address faster? Currently search by stake address takes longer time. Is it because kupo tries to find all owned addresses and find utxos for each address?

**Comment by KtorZ** [comment]
Searching by stake address takes indeed some time, because it is entire addresses that are stored, and not only the stake part. The index is optimized for addresses because that's the most common use-case. It should still remain relatively fast for stake address too however. 

May I ask about how long it takes for you and what version you're using? 

Note: I also just noticed that it seems that the index isn't used for searching addresses by stake.. which would explain why it is slow as it has to scan through the whole database. That's a problem though because the index should be used here too! I'll look into this.

**Reply by baymac** [reply]
Thanks for your reply @KtorZ. We need optimised searching by stake address for 2 use cases:

1. Get balance of account
2. Get UTxOs of account for txn building

If searching by stake address could be as fast as search by address it would be best. Is it possible to add a configuration to store data against the stake address if required? Would it require a major rework?

**Reply by KtorZ** [reply]
> If searching by stake address could be as fast as search by address it would be best

I think that is doable. 
May I ask what's the current time it takes for you, what version are you running and with which options (in particular, with or without `--prune-utxo`?)

**Reply by baymac** [reply]
For a couple of sample accounts:

1. Single Address - takes anywhere between 2-10s (usually 5s)
2. Stake Address - between 35s-1.15mins (usually around 40s, rarely exceeds 1 min)

I am usually master (revision: https://github.com/CardanoSolutions/kupo/commit/3830215cb44b7e95e8940c2374fa08b228fd9ee1) build artefact. I am running kupo without prune-utxo. My command:

```
~/bin/kupo \
--node-socket ~/cardano-node/node.socket \
--node-config ~/cardano-node/configs/mainnet-config.json \ 
--match "*" \
--workdir ~/cardano-node/db \
--since 16588737.4e9bbbb67e3ae262133d94c3da5bffce7b1127fc436e7433b87668dba34c354a
```

**Reply by KtorZ** [reply]
I have a good and bad news. 

The good news is that, I've been able to tweak the address serialization so that the database's index can be leveraged on each query for both search by address and search by stake address. As a result, the queries are now ~50x faster. 

The bad news being that this changes the underlying database structure in a way that can't easily be migrated. Said differently, you'll need to re-synchronize your index.

**Reply by baymac** [reply]
@KtorZ are these source changes or just databases tweaks? What do you mean by resynchronize? Can you tell me the steps to do it?

**Reply by KtorZ** [reply]
```
rm kupo.sqlite3
```

😅... And restart. I haven't yet pushed the changes though. I need to add a few more comments and details explaining what I did before pushing it upstream. It is non-trivial.

I'll let you know once pushed. You'll need to delete your current working directory / database and restart from scratch for the database to be rebuilt.

**Reply by baymac** [reply]
Sounds good, thanks 👍

**Reply by KtorZ** [reply]
Latest master should now perform a lot better when it comes to querying addresses by stake credentials.

**Reply by baymac** [reply]
The changes looks great. Query by stake credentials response is almost instant. Thanks @KtorZ for working on this. :)

**Reply by KtorZ** [reply]
> Query by stake credentials response is almost instant. 

It should be :smile: ! 

> Thanks @KtorZ for working on this. :)

Thanks for raising this _now_. I was actually right in the middle of benchmarks as I am wrapping up the next major release (v2.0.0). So it would have been really unfortunate to learn about this _after the release_ since this required altering the database schema in a non-backward compatible way. So until v2.0.0 I am still fairly open to breaking changes, it's a good opportunity to pack them altogether. 

Cheers!

**Comment by longngn** [comment]
Does SQLite have hash table index so we could reduce query complexity even more from O(log n) to O(1)?

**Reply by KtorZ** [reply]
A hash table would still provide O(log n) access to its elements though :thinking: ? A finger tree could possibly gives amortized O(1), but that'd be to maintain separately and likely in-memory.
